autoload/ale/lsp/response.vim

```diff
@@ -47,7 +47,10 @@ function! ale#lsp#response#ReadDiagnostics(response) abort
         endif
 
         if has_key(l:diagnostic, 'code')
-            let l:loclist_item.nr = l:diagnostic.code
+            let l:loclist_item.code = l:diagnostic.code
+            if type(l:diagnostic.code) == v:t_number
+                let l:loclist_item.nr = l:diagnostic.code
+            endif
```

Comment:
Check the value of the code before you set it. Don't set it if it's `-1`. Convert numbers to strings.

Put blank lines above if statements that don't start a new level of indentation.

Reply:
> Check the value of the code before you set it. Don't set it if it's -1

Will do for `nr`; not planning on doing that for `code` given that `-1` has no special meaning either here (after removing the default), or in the LSP

Reply:
The `code` key should be a String, and shouldn't be set if the code isn't valid.

Reply:
> shouldn't be set if the code isn't valid.

`-1` is a valid code for an LSP server to send; the only requirement in the specification is that it is a number or a string.

> The code key should be a String

Sure, planning on setting it as `'-1'` in this case

Reply:
I wouldn't bother setting the code if it's `-1`. It doesn't tell anyone anything useful. They see a message like `-1: Some error`. It's better if they just see `Some error` instead.

Reply:
Do you know if any servers use `-1` for "no error code," or if any servers use `-1` for a specific error code?

Reply:
Fair enough. That said, both this behavior and casting to strings will likely lead to problems if you decide to implement `textDocument/codeAction` in the future, as most servers seem to expect the code provided by the client in context/diagnostics to exactly match what was sent.

Reply:
We could add another key for that. The existing key named `code` has a specific meaning in ALE, which is an error code to print to the user. This is typically a descriptive code like `'no-const-assign'` or something.